File: app/src/main/java/com/ty/zbpet/ui/activity/material/PickOutDoneDetailActivity.java
package com.ty.zbpet.ui.activity.material;

import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.InputType;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import com.bigkoo.pickerview.listener.OnTimeSelectListener;
import com.ty.zbpet.R;
import com.ty.zbpet.bean.ResponseInfo;
import com.ty.zbpet.bean.material.MaterialDetailsOut;
import com.ty.zbpet.bean.material.MaterialDoneSave;
import com.ty.zbpet.net.HttpMethods;
import com.ty.zbpet.presenter.material.MaterialUiListInterface;
import com.ty.zbpet.presenter.material.PickOutPresenter;
import com.ty.zbpet.ui.adapter.material.PickingDoneDetailAdapter;
import com.ty.zbpet.ui.base.BaseActivity;
import com.ty.zbpet.ui.widght.SpaceItemDecoration;
import com.ty.zbpet.util.CodeConstant;
import com.ty.zbpet.util.DataUtils;
import com.ty.zbpet.util.ResourceUtil;
import com.ty.zbpet.util.ZBUiUtils;
import com.zhouyou.http.exception.ApiException;
import com.zhouyou.http.subsciber.BaseSubscriber;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import okhttp3.RequestBody;

/**
 * @author TY on 2018/11/22.
 * 领料出库 已办详情
 */
public class PickOutDoneDetailActivity extends BaseActivity implements MaterialUiListInterface<MaterialDetailsOut.ListBean> {


    private RecyclerView reView;
    private TextView tvTime;
    private EditText etDesc;
    private Button addButton;

    private PickingDoneDetailAdapter adapter;

    private String selectTime;
    private String sapOrderNo;
    private String orderId;

    private String warehouseId;
    private String mOutWarehouseOrderId;
    private List<MaterialDetailsOut.ListBean> list = new ArrayList<>();


    private PickOutPresenter presenter = new PickOutPresenter(this);


    @Override
    protected void onBaseCreate(Bundle savedInstanceState) {

    }

    @Override
    protected int getActivityLayout() {
        return R.layout.activity_content_row_two;
    }

    @Override
    protected void initOneData() {
        sapOrderNo = getIntent().getStringExtra("sapOrderNo");
        mOutWarehouseOrderId = getIntent().getStringExtra("mOutWarehouseOrderId");

        orderId = getIntent().getStringExtra("orderId");

        presenter.fetchPickOutDoneListDetails(orderId);
    }

    @Override
    protected void initTwoView() {


        initToolBar(R.string.pick_out_storage, new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                pickOutDoneSave(initDoneBody());
            }
        });

        reView = findViewById(R.id.rv_in_storage_detail);
        tvTime = findViewById(R.id.tv_time);
        etDesc = findViewById(R.id.et_desc);
        addButton = findViewById(R.id.add_ship);
        addButton.setVisibility(View.GONE);

        SimpleDateFormat format = new SimpleDateFormat(CodeConstant.DATE_SIMPLE_H_M, Locale.CHINA);
        selectTime = format.format(new Date());

        tvTime.setText(selectTime);

        etDesc.setInputType(InputType.TYPE_NULL);
    }

    /**
     * 出库 保存
     */
    private void pickOutDoneSave(RequestBody body) {

        HttpMethods.getInstance().pickOutDoneSave(new BaseSubscriber<ResponseInfo>() {
            @Override
            public void onError(ApiException e) {
                ZBUiUtils.showToast(e.getMessage());
            }

            @Override
            public void onNext(ResponseInfo responseInfo) {
                if (CodeConstant.SERVICE_SUCCESS.equals(responseInfo.getTag())) {
                    // 入库成功（保存）
                    ZBUiUtils.showToast(responseInfo.getMessage());
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            finish();
                        }
                    });
                } else {
                    ZBUiUtils.showToast(responseInfo.getMessage());
                }
            }
        },body);
    }


    /**
     * 构建保存 body
     * @return
     */
    private RequestBody initDoneBody() {

        MaterialDoneSave requestBody = new MaterialDoneSave();

        requestBody.setOrderId(mOutWarehouseOrderId);
        requestBody.setWarehouseId(warehouseId);
        requestBody.setOutTime(selectTime);
        requestBody.setOrderId(orderId);
        String json = DataUtils.toJson(requestBody, 1);

        return RequestBody.create(okhttp3.MediaType.parse("application/json;charset=UTF-8"), json);
    }

    @Override
    public void showMaterial(List<MaterialDetailsOut.ListBean> list) {
        warehouseId = list.get(0).getWarehouseId();

        if (adapter == null) {
            LinearLayoutManager manager = new LinearLayoutManager(ResourceUtil.getContext());
            reView.addItemDecoration(new SpaceItemDecoration(ResourceUtil.dip2px(10), false));
            reView.setLayoutManager(manager);
            adapter = new PickingDoneDetailAdapter(this, R.layout.item_material_detail_three_done, list);
            reView.setAdapter(adapter);

            adapter.setOnItemClickListener(new PickingDoneDetailAdapter.OnItemClickListener() {
                @Override
                public void onItemClick(View view, RecyclerView.ViewHolder holder, int position) {

                    View rlDetail = holder.itemView.findViewById(R.id.gone_view);
                    ImageView ivArrow = holder.itemView.findViewById(R.id.iv_arrow);

                    if (rlDetail.getVisibility() == View.VISIBLE) {
                        rlDetail.setVisibility(View.GONE);
                        ivArrow.setImageResource(R.mipmap.ic_collapse);
                    } else {
                        rlDetail.setVisibility(View.VISIBLE);
                        ivArrow.setImageResource(R.mipmap.ic_expand);
                    }

                    ZBUiUtils.hideInputWindow(view.getContext(), view);

                }

                @Override
                public boolean onItemLongClick(View view, RecyclerView.ViewHolder holder, int position) {
                    return false;
                }
            });
        } else {
            adapter.notifyDataSetChanged();
        }

    }


    @Override
    public void showLoading() {

    }

    @Override
    public void hideLoading() {

    }
}

File: app/src/main/java/com/ty/zbpet/ui/activity/material/PickOutDoneDetailActivity.kt
package com.ty.zbpet.ui.activity.material

import android.os.Bundle
import androidx.recyclerview.widget.LinearLayoutManager
import android.widget.CheckBox
import android.widget.EditText
import com.ty.zbpet.R
import com.ty.zbpet.bean.CarPositionNoData
import com.ty.zbpet.bean.material.MaterialDetails
import com.ty.zbpet.bean.material.MaterialDoneSave
import com.ty.zbpet.constant.CodeConstant
import com.ty.zbpet.net.RequestBodyJson
import com.ty.zbpet.presenter.material.MaterialUiListInterface
import com.ty.zbpet.presenter.material.PickOutPresenter
import com.ty.zbpet.ui.adapter.material.PickingDoneDetailAdapter
import com.ty.zbpet.base.BaseActivity
import com.ty.zbpet.constant.TipString
import com.ty.zbpet.ui.widght.ShowDialog
import com.ty.zbpet.ui.widght.SpaceItemDecoration
import com.ty.zbpet.util.DataUtils
import com.ty.zbpet.util.ResourceUtil
import com.ty.zbpet.util.SimpleCache
import com.ty.zbpet.util.ZBUiUtils
import com.xiasuhuei321.loadingdialog.view.LoadingDialog
import kotlinx.android.synthetic.main.activity_content_reversal.*
import okhttp3.RequestBody

/**
 * @author TY on 2018/11/22.
 * 领料出库 已办详情
 */
class PickOutDoneDetailActivity : BaseActivity(), MaterialUiListInterface<MaterialDetails.ListBean> {


    private lateinit var adapter: PickingDoneDetailAdapter

    private lateinit var sapOrderNo: String
    private lateinit var orderId: String
    private var materielVoucherNo: String? = ""
    private lateinit var listBean: List<MaterialDetails.ListBean>

    private val presenter = PickOutPresenter(this)

    override val activityLayout: Int
        get() = R.layout.activity_content_reversal


    override fun onBaseCreate(savedInstanceState: Bundle?) {

    }

    override fun initOneData() {
        sapOrderNo = intent.getStringExtra("sapOrderNo")
        orderId = intent.getStringExtra("orderId")
        materielVoucherNo = intent.getStringExtra("materielVoucherNo")

        presenter.fetchPickOutDoneListDetails(orderId)
    }

    override fun initTwoView() {

        initToolBar(R.string.pick_out_storage_reversal)
        bt_reversal.setOnClickListener { view ->
            ZBUiUtils.hideInputWindow(view.context, view)
            initDoneBody()?.let { presenter.pickOutDoneSave(it) }
        }

    }

    /**
     * 构建保存 body
     * @return
     */
    private fun initDoneBody(): RequestBody? {

        val data = MaterialDoneSave()
        val list = ArrayList<MaterialDoneSave.ListBean>()
        val size = listBean.size

        for (i in 0 until size) {

            val view = recycler_reversal.getChildAt(i)
            if (view != null) {
                val checkBox = view.findViewById<CheckBox>(R.id.check)

                val backNum = listBean[i].backNum
                // 已冲销数量
                val num = view.findViewById<EditText>(R.id.et_back_num).text.toString()
                //val value = SimpleCache.getNumber(i.toString())
                if (checkBox.isChecked && num.isNotEmpty()) {
                    val subNumberService = Integer.parseInt(backNum)
                    val subNumber = Integer.parseInt(num)
                    val countNumber = subNumberService + subNumber
                    // 总量
                    val giveNumber = Integer.parseInt(listBean[i].giveNumber)
                    if (giveNumber < countNumber) {
                        ZBUiUtils.showWarning("冲销数量不能大于总量")
                        return null
                    }
                    val bean = MaterialDoneSave.ListBean()
                    bean.id = listBean[i].id
                    bean.backNum = num
                    bean.giveNumber = listBean[i].giveNumber
                    list.add(bean)
                }
            }
        }

        if (list.size == 0) {
            ZBUiUtils.showWarning("请选择您要冲销的列表")
            return null
        }

        data.list = list
        data.orderId = orderId
        data.moveType = "262"

        val json = DataUtils.toJson(data, 1)

        return RequestBodyJson.requestBody(json)
    }

    override fun showMaterial(list: List<MaterialDetails.ListBean>) {
        listBean = list
        if (list.isEmpty()) {
            return
        }
//        for (i in 0 until list.size) {
//            SimpleCache.clearKey(i.toString())
//        }

        val manager = androidx.recyclerview.widget.LinearLayoutManager(ResourceUtil.getContext())
        recycler_reversal.addItemDecoration(SpaceItemDecoration(ResourceUtil.dip2px(CodeConstant.ITEM_DECORATION), false))
        recycler_reversal.layoutManager = manager
        // 禁止 RecycleView 回收
        recycler_reversal.recycledViewPool.setMaxRecycledViews(0, 0)
        adapter = PickingDoneDetailAdapter(this, materielVoucherNo, R.layout.item_reversal_check, list)
        recycler_reversal.adapter = adapter

    }

    override fun showCarSuccess(position: Int, carData: CarPositionNoData?) {

    }

    override fun saveSuccess() {
        ZBUiUtils.showSuccess(TipString.success)
        finish()
    }

    override fun showError(msg: String?) {
        ZBUiUtils.showError(msg)
    }

    private var dialog: LoadingDialog? = null
    override fun showLoading() {
        dialog = ShowDialog.showFullDialog(this@PickOutDoneDetailActivity, TipString.saveIng)
    }

    override fun hideLoading() {
        dialog?.close()
    }
}
